From 836d68ec691fdab35bcff312b430edbb6efd3ae6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Tue, 10 Dec 2024 23:29:07 +0100
Subject: [PATCH 84/86] xmllite: Access memory with right typed pointer.

/home/bernhard/data/entwicklung/2024/wine/wine/tools/runtest -q -P wine -T . -M xmllite.dll -p dlls/xmllite/tests/x86_64-windows/xmllite_test.exe reader && touch dlls/xmllite/tests/x86_64-windows/reader.ok
06a4:fixme:xmllite:xmlreader_QueryInterface interface {0000000c-0000-0000-c000-000000000046} not implemented
=================================================================
==1696==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffffe1ff8c0 at pc 0x6ffff8583f83 bp 0x7ffffe1fed00 sp 0x7ffffe1fed48
WRITE of size 8 at 0x7ffffe1ff8c0 thread T0
06ac:fixme:file:server_get_file_info Unsupported info class e
    #0 0x6ffff8583f82 in xmlreader_GetProperty /home/bernhard/data/entwicklung/2024/wine\wine/dlls/xmllite/reader.c:2807:20
    #1 0x000140002254 in IXmlReader_GetProperty /home/bernhard/data/entwicklung/2024/wine/wine-build/build-asan-pe/64/obj\include\xmllite.h:331:12
    #2 0x000140002254 in test_reader_create /home/bernhard/data/entwicklung/2024/wine\wine/dlls/xmllite/tests/reader.c:582:10
    #3 0x000140002254 in func_reader /home/bernhard/data/entwicklung/2024/wine\wine/dlls/xmllite/tests/reader.c:2676:5
    #4 0x00014006bad1 in run_test /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:765:5
    #5 0x00014006bad1 in main /home/bernhard/data/entwicklung/2024/wine\wine/include/wine/test.h:884:12
    #6 0x00014006d94f in mainCRTStartup /home/bernhard/data/entwicklung/2024/wine\wine/dlls/msvcrt/crt_main.c:58:11
    #7 0x6ffffabf4808 in BaseThreadInitThunk /home/bernhard/data/entwicklung/2024/wine/wine/dlls/kernel32\thread.c:61:5
    #8 0x6ffffadcfa1a in RtlUserThreadStart (C:\windows\system32\ntdll.dll+0x17000fa1a)

Address 0x7ffffe1ff8c0 is located in stack of thread T0 at offset 2784 in frame
    #0 0x00014000100f in func_reader /home/bernhard/data/entwicklung/2024/wine\wine/dlls/xmllite/tests/reader.c:2675

  This frame has 113 object(s):
    [32, 36) 'd.i439.i' (line 824)
    [48, 52) 'd.i412.i' (line 824)
    [64, 68) 'd.i385.i' (line 824)
    [80, 84) 'd.i358.i' (line 824)
    [96, 100) 'd.i307.i' (line 824)
    [112, 116) 'd.i280.i' (line 824)
    [128, 132) 'd.i253.i' (line 824)
    [144, 148) 'd.i178.i' (line 824)
    [160, 164) 'd.i.i2162' (line 824)
    [176, 184) 'reader.i2163' (line 2595)
    [208, 216) 'reader.i1948' (line 2473)
    [240, 248) 'reader.i1791' (line 2407)
    [272, 276) 'type.i1792' (line 2408)
    [288, 292) 'position.i' (line 2409)
    [304, 308) 'd.i320.i' (line 824)
    [320, 324) 'd.i294.i' (line 824)
    [336, 340) 'd.i267.i' (line 824)
    [352, 356) 'd.i240.i' (line 824)
    [368, 372) 'd.i177.i' (line 824)
    [384, 388) 'd.i150.i' (line 824)
    [400, 404) 'd.i.i1643' (line 824)
    [416, 420) 'nodetype.i1644' (line 2311)
    [432, 440) 'reader.i1645' (line 2313)
    [464, 472) 'reader.i1535' (line 2242)
    [496, 500) 'type.i1536' (line 2243)
    [512, 520) 'reader.i1450' (line 2190)
    [544, 548) 'type.i1451' (line 2191)
    [560, 568) 'value.i1315' (line 2137)
    [592, 600) 'reader.i1316' (line 2138)
    [624, 628) 'type.i1317' (line 2139)
    [640, 648) 'reader.i1225' (line 2072)
    [672, 676) 'type.i1226' (line 2073)
    [688, 696) 'local.i1227' (line 2087)
    [720, 728) 'qname.i1228' (line 2087)
    [752, 756) 'length.i' (line 2088)
    [768, 772) 'length2.i' (line 2088)
    [784, 792) 'reader.i1156' (line 1998)
    [816, 820) 'type.i1157' (line 2006)
    [832, 840) 'reader.i1102' (line 1959)
    [864, 872) 'value.i1103' (line 1960)
    [896, 900) 'd.i723.i' (line 824)
    [912, 916) 'd.i473.i' (line 824)
    [928, 932) 'd.i.i' (line 824)
    [944, 952) 'reader.i985' (line 849)
    [976, 980) 'type.i986' (line 852)
    [992, 996) 'count.i987' (line 853)
    [1008, 1012) 'len.i988' (line 853)
    [1024, 1032) 'val.i' (line 855)
    [1056, 1064) 'reader.i891' (line 1600)
    [1088, 1092) 'type.i892' (line 1601)
    [1104, 1232) 'buf.i' (line 1602)
    [1264, 1266) 'b.i' (line 1603)
    [1280, 1284) 'c.i' (line 1605)
    [1296, 1304) 'reader.i812' (line 1564)
    [1328, 1336) 'value.i813' (line 1565)
    [1360, 1364) 'type.i814' (line 1566)
    [1376, 1384) 'reader.i795' (line 1254)
    [1408, 1416) 'reader.i668' (line 1763)
    [1440, 1444) 'type.i669' (line 1771)
    [1456, 1464) 'str.i670' (line 1787)
    [1488, 1492) 'len.i671' (line 1788)
    [1504, 1512) 'reader.i564' (line 1835)
    [1536, 1540) 'type.i565' (line 1836)
    [1552, 1560) 'reader.i469' (line 1417)
    [1584, 1588) 'type.i470' (line 1418)
    [1600, 1604) 'depth.i' (line 1420)
    [1616, 1624) 'str.i471' (line 1439)
    [1648, 1652) 'len.i472' (line 1440)
    [1664, 1668) 'count.i473' (line 1464)
    [1680, 1688) 'prefix.i' (line 1476)
    [1712, 1720) 'local.i' (line 1485)
    [1744, 1752) 'qname.i' (line 1485)
    [1776, 1784) 'value.i474' (line 1518)
    [1808, 1816) 'reader.i350' (line 1278)
    [1840, 1848) 'str.i351' (line 1279)
    [1872, 1876) 'type.i352' (line 1280)
    [1888, 1892) 'len.i353' (line 1281)
    [1904, 1908) 'count.i354' (line 1281)
    [1920, 1928) 'reader.i302' (line 1340)
    [1952, 1960) 'str.i303' (line 1341)
    [1984, 1988) 'type.i304' (line 1342)
    [2000, 2004) 'len.i305' (line 1343)
    [2016, 2020) 'count.i' (line 1343)
    [2032, 2040) 'reader.i232' (line 1159)
    [2064, 2068) 'type.i233' (line 1167)
    [2080, 2088) 'str.i234' (line 1179)
    [2112, 2116) 'len.i235' (line 1180)
    [2128, 2136) 'reader.i146' (line 1077)
    [2160, 2164) 'type.i147' (line 1078)
    [2176, 2184) 'value.i148' (line 1088)
    [2208, 2216) 'str.i149' (line 1113)
    [2240, 2248) 'reader.i100' (line 1689)
    [2272, 2276) 'type.i101' (line 1697)
    [2288, 2296) 'str.i102' (line 1717)
    [2320, 2324) 'len.i103' (line 1718)
    [2336, 2344) 'reader.i70' (line 1906)
    [2368, 2372) 'type.i' (line 1914)
    [2384, 2392) 'str.i' (line 1926)
    [2416, 2420) 'len.i' (line 1927)
    [2432, 2436) 'nodetype.i49' (line 780)
    [2448, 2456) 'reader.i50' (line 781)
    [2480, 2488) 'reader_input.i' (line 616)
    [2512, 2520) 'reader.i1' (line 617)
    [2544, 2552) 'reader2.i' (line 617)
    [2576, 2584) 'obj.i' (line 618)
    [2608, 2616) 'stream.i' (line 619)
    [2640, 2648) 'stream2.i' (line 619)
    [2672, 2676) 'nodetype.i2' (line 620)
    [2688, 2696) 'resolver.i' (line 515)
    [2720, 2728) 'unk.i' (line 516)
    [2752, 2760) 'reader.i' (line 517)
    [2784, 2788) 'dtd.i' (line 518) <== Memory access at offset 2784 partially overflows this variable
    [2800, 2804) 'nodetype.i' (line 519)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp, SEH and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /home/bernhard/data/entwicklung/2024/wine\wine/dlls/xmllite/reader.c:2807:20 in xmlreader_GetProperty
Shadow bytes around the buggy address:
  0x7ffffe1ff600: f8 f2 f2 f2 f8 f2 f8 f2 f2 f2 f8 f2 f8 f2 f2 f2
  0x7ffffe1ff680: f8 f2 f2 f2 f8 f2 f2 f2 f8 f2 f8 f2 f2 f2 f8 f2
  0x7ffffe1ff700: f8 f2 f2 f2 f8 f2 f8 f2 f2 f2 f8 f2 f8 f2 f8 f2
  0x7ffffe1ff780: f2 f2 f8 f2 f2 f2 f8 f2 f2 f2 f8 f2 f2 f2 f8 f2
  0x7ffffe1ff800: f2 f2 f8 f2 f2 f2 f8 f2 f2 f2 f8 f2 00 f2 f2 f2
=>0x7ffffe1ff880: 00 f2 f2 f2 00 f2 f2 f2[04]f2 04 f3 00 00 00 00
  0x7ffffe1ff900: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffffe1ff980: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffffe1ffa00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffffe1ffa80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffffe1ffb00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1696==ABORTING
06ac:fixme:kernelbase:AppPolicyGetProcessTerminationMethod FFFFFFFFFFFFFFFA, 00007FFFFEA8FE80
make: *** [Makefile:471569: dlls/xmllite/tests/x86_64-windows/reader.ok] Fehler 1
---
 dlls/xmllite/reader.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/xmllite/reader.c b/dlls/xmllite/reader.c
index 1e146493aa6..583a0fa600c 100644
--- a/dlls/xmllite/reader.c
+++ b/dlls/xmllite/reader.c
@@ -2804,7 +2804,7 @@ static HRESULT WINAPI xmlreader_GetProperty(IXmlReader* iface, UINT property, LO
                 IXmlResolver_AddRef(This->resolver);
             break;
         case XmlReaderProperty_DtdProcessing:
-            *value = This->dtdmode;
+            *(DtdProcessing*)value = This->dtdmode;
             break;
         case XmlReaderProperty_ReadState:
             *value = This->state;
-- 
2.47.1

